BUG: df.replace over pd.Period columns (#34871)

samc1213 opened the pull request:
- [x] closes #34871
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This commit ensures that PeriodArrays return False for _can_hold_element for any element that is not a pd.Period. This prevents upstream code from casting the dtype to object. Also un-xfail test written in #34904

samc1213 committed:
BUG: df.replace over pd.Period columns (#34871)

samc1213 commented:
This is my first PR, so I would really appreciate guidance on how to improve the commit, and/or feedback on any process I am not following properly. Thanks in advance

jbrockmendel commented on pandas/core/internals/blocks.py:
```diff
@@ -1975,6 +1976,18 @@ def external_values(self):
+    Used by PeriodArray to ensure proper type conversions
+    """
+
+    def _can_hold_element(self, element: Any) -> bool:
```
this looks good, but is there any way to do this within ObjectValuesExtensionBlock?  (big picture we're looking to get less special-casing of Block subclasses)

so maybe in ObjectValuesExtensionBlock could write

```
def _can_hold_element(self, element):
        tipo = maybe_infer_dtype_type(element)
        if tipo is not None:
            return issubclass(tipo.type, self.dtype.type)
        return isinstance(element, self.dtype.type)
```

?  Might need to check/test that this works for IntervalDtype too (wouldnt surprise me if we see the analogous bug there)

jbrockmendel commented on pandas/core/internals/blocks.py:
```diff
@@ -10,6 +10,7 @@
 import pandas._libs.internals as libinternals
 from pandas._libs.internals import BlockPlacement
 from pandas._libs.tslibs import conversion
+from pandas._libs.tslibs.period import Period
```
let's just fold this import into the previous line and get it from tslibs

jbrockmendel commented:
> This is my first PR, so I would really appreciate guidance on how to improve the commit, and/or feedback on any process I am not following properly. Thanks in advance

You're doing well so far.  A couple of comments, then the next thing to do will be to add a note in doc/source/whatsnew/1.2.0.rst giving a short description of the fixed bug

jbrockmendel commented on pandas/tests/frame/methods/test_replace.py:
```diff
@@ -1517,9 +1517,6 @@ def test_replace_with_duplicate_columns(self, replacement):
-    @pytest.mark.xfail(
-        reason="replace() changes dtype from period to object, see GH34871", strict=True
-    )
     def test_replace_period_ignore_float(self):
```
does this also affect/fix Series.replace?  if so let's make sure we have a test for that too

samc1213 commented:
> > This is my first PR, so I would really appreciate guidance on how to improve the commit, and/or feedback on any process I am not following properly. Thanks in advance
> 
> You're doing well so far. A couple of comments, then the next thing to do will be to add a note in doc/source/whatsnew/1.2.0.rst giving a short description of the fixed bug

Thanks for your feedback, addressing your comments now. Also, what is the cause of the build failures? They seem to be happening to everyone, is someone working on fixing them? I was trying to find documentation of the build issue, if it is in fact a widespread thing.

samc1213 committed:
Remove PeriodExtensionBlock, update docs/tests

pep8speaks commented:
Hello @samc1213! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`pandas/tests/series/methods/test_replace.py`](https://github.com/pandas-dev/pandas/blob/7b40ecef957be30651d92cb2be64e30aa87c9e66/pandas/tests/series/methods/test_replace.py):

> [Line 456:89](https://github.com/pandas-dev/pandas/blob/7b40ecef957be30651d92cb2be64e30aa87c9e66/pandas/tests/series/methods/test_replace.py#L456): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 88 characters)
> [Line 462:49](https://github.com/pandas-dev/pandas/blob/7b40ecef957be30651d92cb2be64e30aa87c9e66/pandas/tests/series/methods/test_replace.py#L462): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

samc1213 committed:
Fix PEP violations

samc1213 committed:
Fix docs: method to meth

jbrockmendel commented:
> Also, what is the cause of the build failures? They seem to be happening to everyone, is someone working on fixing them? 

Yes, don't worry about it (already fixed in fact)

samc1213 commented:
Pardon my ignorance, is there anything else I need to address/do to get this PR pulled? Thanks for all your help @jbrockmendel

jreback reviewed (changes requested):
can u run some asvs (indexing / reshaping ones)

this could potentially cause some perf issues

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -387,6 +387,11 @@ I/O
+Period
+^^^^^^
+
+- Bug in :meth:`DataFrame.replace` and :meth:`Series.replace` where :class:`Period` dtypes would be converted to object dytpes (:issue:34871)
```
typo on dtypes

jbrockmendel replied:
> dytpes

typo is still there.  BTW in general let the person who made the original comment hit the "conversation resolved" button

samc1213 replied:
Gotcha thanks for the heads up, haven't used github at all really. Fixed

jreback commented on pandas/tests/frame/methods/test_replace.py:
```diff
@@ -1517,17 +1517,15 @@ def test_replace_with_duplicate_columns(self, replacement):
-    )
-    def test_replace_period_ignore_float(self):
+    @pytest.mark.parametrize("value", [pd.Period("2020-01"), pd.Interval(0, 5)])
+    def test_replace_period_ignore_float(self, value):
```
can you change the text name itself (period -> EA)

jreback commented on pandas/tests/series/methods/test_replace.py:
```diff
@@ -449,3 +449,15 @@ def test_replace_with_compiled_regex(self):
         tm.assert_series_equal(result, expected)
+
+    @pytest.mark.parametrize("value", [pd.Period("2020-01"), pd.Interval(0, 5)])
+    def test_replace_period_ignore_float(self, value):
```
same here

jreback reviewed (changes requested):
also there was a reference in the OP to https://github.com/pandas-dev/pandas/issues/35268, if this fixes as well would be great to add the test (if not ok too)

samc1213 commented:
> can u run some asvs (indexing / reshaping ones)
> 
> this could potentially cause some perf issues

Is there a server to run these on? Having issues running these locally in my docker env

jbrockmendel commented:
> Is there a server to run these on?

No.

> Having issues running these locally in my docker env

are you familiar with using ipython's `%timeit`?

jreback commented:
can you merge master

samc1213 commented:
> can you merge master

Will do, sorry I've taken so long on this, haven't had much time to try and get the ASVs working

samc1213 committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into 34871

samc1213 committed:
Merge remote-tracking branch 'upstream/master' into 34871

samc1213 committed:
Merge remote-tracking branch 'upstream/master' into 34871

samc1213 committed:
Address test and release notes comments

samc1213 commented:
> also there was a reference in the OP to #35268, if this fixes as well would be great to add the test (if not ok too)

It does not fix #35268 - however, I think it's a very similar code change to fix - I can take care of that once this is approved

samc1213 commented:
> > Is there a server to run these on?
> 
> No.
> 
> > Having issues running these locally in my docker env
> 
> are you familiar with using ipython's `%timeit`?

I think I've got this working, will reply with results shortly

samc1213 commented:
> > also there was a reference in the OP to #35268, if this fixes as well would be great to add the test (if not ok too)
> 
> It does not fix #35268 - however, I think it's a very similar code change to fix - I can take care of that once this is approved

Actually, I will modify this branch to fix both, should be a small change

samc1213 committed:
Fix unit tests

samc1213 commented:
> > > Is there a server to run these on?
> > 
> > 
> > No.
> > > Having issues running these locally in my docker env
> > 
> > 
> > are you familiar with using ipython's `%timeit`?
> 
> I think I've got this working, will reply with results shortly

![image](https://user-images.githubusercontent.com/9583514/97095983-f71dca80-1633-11eb-95db-e091bce00321.png)

Looks like I'm getting some bad results for indexing ASVs. I'm not sure why this is affecting the NumericSeriesIndexing tests, this shouldn't be affecting numeric series?

samc1213 committed:
More build fixes

jbrockmendel commented:
> Looks like I'm getting some bad results for indexing ASVs. I'm not sure why this is affecting the NumericSeriesIndexing tests, this shouldn't be affecting numeric series?

Unfortunately its not unusual for the asvs to include some noise.  Are these results consistent across runs?  (just re-run the surprising outliers)

samc1213 commented:
> > Looks like I'm getting some bad results for indexing ASVs. I'm not sure why this is affecting the NumericSeriesIndexing tests, this shouldn't be affecting numeric series?
> 
> Unfortunately its not unusual for the asvs to include some noise. Are these results consistent across runs? (just re-run the surprising outliers)

Thanks, looks like I'm making it faster now 😄 

![image](https://user-images.githubusercontent.com/9583514/97385931-27ba6a00-18a9-11eb-926b-597625b60e07.png)

jreback committed:
Merge branch 'master' into 34871

jreback commented on pandas/core/internals/blocks.py:
```diff
@@ -2002,6 +2002,18 @@ class ObjectValuesExtensionBlock(ExtensionBlock):
+    def _can_hold_element(self, element: Any) -> bool:
+        if is_valid_nat_for_dtype(element, self.dtype):
+            return True
+        if element is NaT:
```
@jbrockmendel ok here?

@samc1213 are all of these branches actually hit? in particular why L2008 and L2010

samc1213 replied:
@jreback Added these due to unit tests. Specifically, L2010 is required for `pandas/tests/extension/test_interval.py::TestSetitem::test_setitem_empty_indxer` to pass. L2008 may not be required, I just removed it and will see if the tests pass...

samc1213 committed:
Remove excess NaT check

jreback commented:
lgtm. @jbrockmendel if you'd look and merge if ok.

jreback commented:
cc @jbrockmendel ok here

jbrockmendel commented:
will take a look this afternoon

jbrockmendel commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -434,6 +434,8 @@ Strings
 Interval
 ^^^^^^^^
+
+- Bug in :meth:`DataFrame.replace` and :meth:`Series.replace` where :class:`Interval` dtypes would be converted to object dtypes (:issue:34871)
```
backticks around the issue number

jbrockmendel replied:
same below

samc1213 replied:
Thanks, fixed

jbrockmendel commented on pandas/core/internals/blocks.py:
```diff
@@ -2002,6 +2002,16 @@ class ObjectValuesExtensionBlock(ExtensionBlock):
+        if is_valid_nat_for_dtype(element, self.dtype):
+            return True
+        if isinstance(element, list) and len(element) == 0:
+            return True
```
why this? i dont think we have this in other _can_hold_element methods

samc1213 replied:
Without it, `pandas/tests/extension/test_interval.py::TestSetitem::test_setitem_empty_indexer` will not pass. It is also used on L2223. If there's a better way to handle this please let me know, but I did borrow this from 2223

jbrockmendel commented on pandas/tests/series/methods/test_replace.py:
```diff
@@ -450,6 +450,18 @@ def test_replace_with_compiled_regex(self):
+        Regression test for corrolary to GH#34871: if series.replace(1.0, 0.0)
+        is called on a Period/Interval Series, the old, faulty behavior
+        is to raise TypeError.
+        """
```
this can be a comment, not a docstring, just needs to see `# GH#34871`, ditto above

jbrockmendel replied:
actually you can share this test using the frame_or_series fixture

samc1213 replied:
Changed to a comment, but I grepped for frame_or_series and not finding much in the codebase. I did find `index_or_series`. Can you please elaborate more on how to use frame_or_series? There do seem to be a fair amount of duplication between these two files (series/test_replace and frame/test_replace). `test_replace_with_compiled_regex` is basically copy-pasted between the two, for example

arw2019 replied:
> Can you please elaborate more on how to use frame_or_series? 

It's a pytest fixture:
https://github.com/pandas-dev/pandas/blob/aa390ecc825464953f91e9214ed65ef5f83be5d6/pandas/conftest.py#L296

samc1213 replied:
Thanks @arw2019  - I needed to merge master in locally to find this

samc1213 committed:
Address review comments

samc1213 committed:
Merge remote-tracking branch 'upstream/master' into 34871

samc1213 committed:
Combine test_replace Series and DataFrame tests

samc1213 committed:
Add missing file

samc1213 committed:
Remove unused import

samc1213 committed:
Newline at end of file

jbrockmendel commented on pandas/tests/generic/methods/test_replace.py:
```diff
@@ -0,0 +1,24 @@
+        regex = re.compile("^a$")
+        result = s.replace({regex: "z"}, regex=True)
+        expected = pd.Series(["z", "b", "c"])
+        tm.assert_series_equal(result, expected)
```
tests only belong in tests.generic.methods if _all_ tests for that method are fully parametrized over DataFrame/Series, which these are not.

(i think this is confusing, so will likely just get rid of this directory at some point

samc1213 replied:
I see, makes sense then. I'll just test frame and series in the same file then

jbrockmendel commented on pandas/tests/frame/methods/test_replace.py:
```diff
@@ -1610,14 +1597,6 @@ def test_replace_dict_category_type(self, input_category_df, expected_category_d
 
         tm.assert_frame_equal(result, expected)
 
-    def test_replace_with_compiled_regex(self):
```
this is unrelated to the current PR, right?  if you want to make a separate PR to parametrize this test, go for it.  better to just use frame_or_series in this file and remove the version in the series tests

samc1213 replied:
Sorry, yeah its unrelated. Undone, and just used frame_or_series here, just trying to organize it properly (evidently was misguided :D )

jbrockmendel replied:
> evidently was misguided

not at all, organizing/parametrizing/breaking-large-ancient-tests-into-specific-tests is something we very much want to encourage.  just in dedicated PRs

samc1213 replied:
I see, definitely makes sense - thanks

samc1213 committed:
Remove generic unit test

samc1213 committed:
Add back series regex test

jbrockmendel commented on pandas/tests/frame/methods/test_replace.py:
```diff
@@ -1517,18 +1517,13 @@ def test_replace_with_duplicate_columns(self, replacement):
+    def test_replace_ea_ignore_float(self, frame_or_series, value):
+        # GH#34871
+        df = frame_or_series([value] * 3)
         result = df.replace(1.0, 0.0)
```
since this is now frame_or_series, can you call it `obj` instead of `df`

samc1213 replied:
Got it, done

jbrockmendel commented:
One last nitpick, otherwise LGTM, cc @jreback

samc1213 committed:
Nitpick df to obj

samc1213 committed:
Merge remote-tracking branch 'upstream/master' into 34871

samc1213 committed:
Fix conflict

jreback commented:
thanks @samc1213

samc1213 commented:
@jreback  @jbrockmendel Thanks so much for your patience and feedback, this was a cool experience for me. Hopefully I'll be back! Appreciate all the time you put into this great project
